### DISCUSSION TITLE
Can stats be opened via macro?

### DISCUSSION THREAD

**Discussion by JornieNonsubia** [jacobwojoski/dice-stats#143] [Q&A] [answered]
Hi, in my game for more imersive experience I hide almost all UI for players, but I want them to be able to access dice stats using the landing page I've made. 
Is there any possibility to open it via macro?

**Comment by jacobwojoski** [comment]
Should be possible. I’ll see if I can write up some sort of macro to do it for ya. it will probably a macro that opens a display with all the same buttons & names.

**Comment by jacobwojoski** [comment]
Well, looks like there's not a great way to do it. It looks like I can't get the values from my mod from a macro. Can't find a way to access module data from macros in general

**Reply by jacobwojoski** [reply]
If anyone wants to try working from what I have here what should work. Just need a way to access the global module object. Im guessing id need to save data to the flags vars to get it to work tho.

`
// import { DS_GLOBALS } from "modules/diceStats/scripts/dice-stats-globals.js";
// import { GlobalStatusPage } from "modules/diceStats/scripts/forms/dice-stats-globalstatuspage.js";
// import { ComparePlayerStatusPage } from "modules/diceStats/scripts/forms/dice-stats-compareplayerspage.js";
// import { ExportDataPage } from "modules/diceStats/scripts/forms/dice-stats-exportdatapage.js";
// import { CustomTabFormClass } from "modules/diceStats/scripts/forms/dice-stats-tabedplayerstatspage.js";

/* Open the global status page */
async function openGlobalStatusPage(){
    //Make sure Compare Is Closed B4 Opening
    await DS_GLOBALS.FORM_GL_COMPARE?.close();

    if(DS_GLOBALS.FORM_GL_STATS){
        DS_GLOBALS.FORM_GL_STATS.render(true);
    }else{
        DS_GLOBALS.FORM_GL_STATS = new GlobalStatusPage().render(true);
    } 
}

/* Open the compare player status page*/
async function openComparePlayerStatusPage () {
    let canSeePlayerData = game.settings.get(DS_GLOBALS.MODULE_ID, DS_GLOBALS.MODULE_SETTINGS.PLAYERS_SEE_PLAYERS);
    if(canSeePlayerData === false && game.user.isGM === false){
        //Do nothing, Dont allow players to view player data if setting is set
        ui.notifications.warn("No Accesss to Player Data, Ask GM For Permission");
    }else if(DS_GLOBALS.FORM_GL_COMPARE){
        // Close Global Stats page if opening Compare popup
        await DS_GLOBALS?.FORM_GL_STATS?.close(false);

        DS_GLOBALS.FORM_GL_COMPARE.render(true);
    }else{
        // Close Global Stats page if opening Compare popup
        await DS_GLOBALS?.FORM_GL_STATS?.close(false);

        DS_GLOBALS.FORM_GL_COMPARE = new ComparePlayerStatusPage().render(true);
    } 
}

/* open player status page */
function openPlayerStatusPage(event){
    const clickedElement = $(event.currentTarget);
    const playerID = clickedElement.data().value;

    let canSeeGM = game.settings.get(DS_GLOBALS.MODULE_ID,DS_GLOBALS.MODULE_SETTINGS.PLAYERS_SEE_GM);
    let canSeePlayerData = game.settings.get(DS_GLOBALS.MODULE_ID,DS_GLOBALS.MODULE_SETTINGS.PLAYERS_SEE_PLAYERS);
    let amIGM = game.users.get(game.userId)?.isGM;
    let isThisGMrolls = game.users.get(playerID)?.isGM;

    if(canSeeGM === false && isThisGMrolls && !amIGM){
        //do nothing, Dont allow ability to see gm data if setting is off
        ui.notifications.warn("No Accesss to GM Data, Ask GM For Permission");
    }else if(canSeePlayerData === false && game.user.isGM === false){
        //Do nothing, Dont allow players to view player data if setting is set
        ui.notifications.warn("No Accesss to Player Data, Ask GM For Permission");
    }else{
        
        //Delete OLD OBJ before making new one (Other objects are singletons so delete is not needed)
        if(DS_GLOBALS.FORM_PLAYER_STATS)
        {
            DS_GLOBALS.FORM_PLAYER_STATS.render(false);
            delete DS_GLOBALS.FORM_PLAYER_STATS
        }

        // Always render the Tabbed Version, Were removing the normal Player status page
        DS_GLOBALS.FORM_PLAYER_STATS = new CustomTabFormClass(playerID).render(true);
    }
}

/* Open the export page */
function openExportDataPage(){
    if (game.user.isGM) {
        DS_GLOBALS.FORM_EXPORT = new ExportDataPage().render(true);
    }else{
        ui.notifications.warn("No Accesss to Player Data, Ask GM For Permission");
    }
}

// ============================================================================
// =================================== MAIN ===================================
// ============================================================================

// game.modules.get("dice-stats")?.value; 
let buttons_list = "";
let player_count = 0;

/* THIS DOESNT WORK FOR SOME REASON */
for ( let user in game.users ){
    buttons_list+='<button id="open_player" data-value='+user.id+' style="padding: 10px 20px;">'+user.name+" : "+user.id+'</button>';
}

// Define the HTML content for the pop-up dialog
let content = `
  <div style="text-align:center; padding: 10px;">
    <p>Click the button below to perform an action:</p>
    <button id="open_global" style="padding: 10px 20px;">Globals</button>
    <button id="open_compare" style="padding: 10px 20px;">Campare</button>
    <button id="open_export" style="padding: 10px 20px;">Export</button>
  ` + buttons_list + 
  `</div>`;

// Create a simple dialog with the HTML content
new Dialog({
  title: "Action Dialog",  // Title of the dialog
  content: content,        // HTML content of the dialog
  buttons: {},             // No other buttons for now
  default: "cancel",       // Default button action
  close: () => {
    // Close action (if needed)
    console.log("Dialog closed");

    // Add event listener for the button click
    $(document).off('click', '#open_global');
    $(document).off('click', '#open_compare');
    $(document).off('click', '#open_export');
    $(document).off('click', '#open_player');
  }
}).render(true);

// Add event listener for the button click
$(document).on('click', '#open_global', (event) => {openGlobalStatusPage(event) });
$(document).on('click', '#open_compare', (event) => {openComparePlayerStatusPage(event) });
$(document).on('click', '#open_export', (event) => {openExportDataPage(event) });
$(document).on('click', '#open_player', (event) => {openPlayerStatusPage(event) });

`

**Comment by jacobwojoski** [comment]
jk, Making an API and it should be able to work. Conna take a bit tho

**Reply by jacobwojoski** [reply]
https://foundryvtt.wiki/en/development/library-modules

**Comment by JornieNonsubia** [comment]
You got me xD 
thank you very much for your help

**Comment by jacobwojoski** [comment, marked as answer]
1.21.0 release has macro support. You should be able to copy paste the macro from [script/utils/open-displays-macro.js](https://github.com/jacobwojoski/dice-stats/blob/main/scripts/utils/open-displays-macro.js) and paste it to be used by GM or players and it should just work.

LMK if you have any problems.
